contracts/oracle/BaseOraclePTPendle.sol

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity ^0.8.12;
+
+import { UNIT, UD60x18, ud } from "prb/math/UD60x18.sol";
+import "pendle/interfaces/IPMarket.sol";
+import { PendlePtOracleLib } from "pendle/oracles/PendlePtOracleLib.sol";
+import "../interfaces/ITreasury.sol";
+
+/// @title BaseOraclePTPendle
+/// @author Angle Labs, Inc.
+/// @notice Base oracle implementation for PT tokens on Pendle
+abstract contract BaseOraclePTPendle {
+    uint256 public constant BASE_18 = 1 ether;
+    uint256 public constant YEAR = 365 days;
+    // @notice The maximum implied rate for the underlying asset, if set well it allows to have a lower bound on the PT token price
+    uint256 public maxImpliedRate;
+    // @notice The duration of the TWAP used to calculate the PT price
+    uint32 public twapDuration;
+
+    /*//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+                                                        EVENTS                                                      
+    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////*/
+
+    event MaxImpliedRateUpdated(uint256 _maxImpliedRate);
+    event TwapPTDurationUpdated(uint256 _twapDuration);
+
+    constructor(uint256 _maxImpliedRate, uint32 _twapDuration) {
+        maxImpliedRate = _maxImpliedRate;
+        twapDuration = _twapDuration;
+    }
+
+    modifier onlyGovernorOrGuardian() {
+        _onlyGovernorOrGuardian();
+        _;
+    }
+
+    function _getQuoteAmount() internal view virtual returns (uint256) {
+        uint256 economicalLowerBound = _economicalPTLowerBoundPrice();
+        uint256 pendlePrice = _pendlePTPrice();
+        uint256 minPrice = economicalLowerBound > pendlePrice ? pendlePrice : economicalLowerBound;
+        uint256 quote = (_detectHackRatio() * minPrice) / BASE_18;
```

Comment:
Note to myself. In case of a hack the collateral price would be lowered pro rata of the missing underlying tokens.